ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala

```diff
@@ -80,12 +80,39 @@ private[platform] class MutableCacheBackedContractStore(
       .flatMap {
         case (cached, toBeFetched) if toBeFetched.isEmpty =>
           Future.successful(Some(cached.max))
+
         case (cached, toBeFetched) =>
-          contractsReader
-            .lookupMaximumLedgerTime(toBeFetched)
-            .map(_.map(m => (cached + m).max))
+          readThroughMaximumLedgerTime(toBeFetched.toList, cached.maxOption)
       }
 
+  private def readThroughMaximumLedgerTime(
+      missing: List[ContractId],
+      acc: Option[Timestamp],
+  ): Future[Option[Timestamp]] =
+    missing match {
+      case contractId :: restOfMissing =>
+        readThroughContractsCache(contractId).flatMap {
+          case active: Active =>
+            val newMaximumLedgerTime = Some(
+              (active.createLedgerEffectiveTime :: acc.toList).max
+            )
+            readThroughMaximumLedgerTime(restOfMissing, newMaximumLedgerTime)
+
+          case _: Archived =>
+            Future.failed(MissingContracts(Set(contractId)))
+
+          case NotFound =>
+            // If cannot be found: no create or archive event for the contract.
+            // Since this contract is part of the input, it was able to be looked up once.
```

Comment:
Dear reviewers, pls let me know what you think about this aspect. This approach only works if we can rely on the fact that the divulged contract was being able to be looked up once.
We use this functionality in two places:
* Command execution on LAPI write side: there we load/look up all input contracts as part of the execution
* sandbox-on-x conflict checking with committed: since we check only tx-es here which were executed before, we defer to the above :point_up:

Reply:
I think this optimization is sane as long as it is coupled with interpretation. I wouldn't see a scenario where they would be decoupled, so I'd say the comment gives enough safety wrt to further development work that might impact this.

Reply:
@nmarton-da : I agree that this should be safe. I'd appreciate if you could add a corresponding comment to the public entry point, which is `ContractStore.lookupMaximumLedgerTime`. Also consider changing it to `lookupMaximumLedgerTimeAfterInterpretation` just to make it doubly clear that its usage is very context-specific.

Reply:
One potential snag that is in principle possible is that the `NotFound` is due to (very) aggressive pruning. The proper safety measure for this is to check the offset as-of-which interpretation began against the `pruningEnd` just before submitting the `tx` to the backing ledger.

It would be good to create an issue for that; perhaps together with the work for incremental pruning?

Reply:
> It would be good to create an issue for that; perhaps together with the work for incremental pruning?

Oh my I am so looking forward to see incremental pruning in action! Would make things so much more proper!
Great point, will add it somewhere (first I have to figure where)

Reply:
>  I agree that this should be safe. I'd appreciate if you could add a corresponding comment to the public entry point, which is ContractStore.lookupMaximumLedgerTime. Also consider changing it to lookupMaximumLedgerTimeAfterInterpretation just to make it doubly clear that its usage is very context-specific.

I will implement this in the next PR, because I plan to change the return type, client code as well. (keeping the original purpose of functional change here)